src/libasr/pass/intrinsic_array_function_registry.h

```diff
@@ -501,72 +533,77 @@ static inline ASR::asr_t* create_ArrIntrinsic(
     Allocator& al, const Location& loc, Vec<ASR::expr_t*>& args,
     diag::Diagnostics& diag, ASRUtils::IntrinsicArrayFunctions intrinsic_func_id) {
     std::string intrinsic_func_name = ASRUtils::get_array_intrinsic_name(static_cast<int64_t>(intrinsic_func_id));
-    int64_t id_array = 0, id_array_dim = 1, id_array_mask = 2;
-    int64_t id_array_dim_mask = 3;
+    int64_t id_array = 0, id_array_dim = 1, id_array_mask = 2, id_array_dim_mask = 3;
     int64_t overload_id = id_array;
 
     ASR::expr_t* array = args[0];
-    ASR::expr_t *arg2 = nullptr, *arg3 = nullptr;
-    if( args.size() >= 2 ) {
-        arg2 = args[1];
-    }
-    if( args.size() == 3 ) {
-        arg3 = args[2];
+    ASR::expr_t *dim = nullptr;
+    ASR::expr_t *mask = nullptr;
+    ASR::ttype_t* array_type = ASRUtils::expr_type(array);
+    if (args[1]) {
+        if (is_integer(*ASRUtils::expr_type(args[1]))) {
+            dim = args[1];
+            if (args[2]) {
+                mask = args[2];
+            }
+        } else if (is_logical(*ASRUtils::expr_type(args[1]))) {
+            mask = args[1];
+            if (args[2]) {
+                dim = args[2];
+            }
+        }
+    } else if (args[2]) {
```

Comment:
```suggestion
    }
    if (args[2]) {
```
should it not be this?

Reply:
If args[1] is present, it checks for args[2] too. So `else if` works in this case.

Reply:
How is this possible, if `args[2]` is present then `args[1]` should definitely right?

Reply:
If this is the case we may not need either of the checks

Reply:
I think no. It may happen that mask is provided and dim is not and vice versa. Any of the two can be null.

Reply:
Then either of the two provided will be `args[1]`

Reply:
Actually, they are handled in a different way for eval and instantiate I believe. I tried what your're suggesting but that led to errors too. So I found the current way to be the best possible.

Reply:
Well I think we need it, I asked chatGPT to club both and this is what it gave me, both seems equivalent, so what you have is fine.

```cpp
if (args[1] || args[2]) {
    if (args[1] && is_integer(*ASRUtils::expr_type(args[1]))) {
        dim = args[1];
        if (args[2]) {
            mask = args[2];
        }
    } else if (args[1] && is_logical(*ASRUtils::expr_type(args[1]))) {
        mask = args[1];
        if (args[2]) {
            dim = args[2];
        }
    } else if (args[2] && is_integer(*ASRUtils::expr_type(args[2]))) {
        dim = args[2];
        mask = args[1];
    } else if (args[2] && is_logical(*ASRUtils::expr_type(args[2]))) {
        mask = args[2];
        dim = args[1];
    }
}
```